src/docmanager/index.ts

```diff
@@ -0,0 +1,958 @@
+    if (id) {
+      return id;
+    } else {
+      return this._contextManager.createNew(path, model, mFactoryEx, contents);
+    }
+  }
+
+  /**
+   * Create a widget from a context and attach it to the parent.
+   */
+  private _createWidget(contextId: string, widgetName: string, parent: Widget, kernel?: IKernelId): void {
+    let wFactoryEx = this._getWidgetFactoryEx(widgetName);
+    if (!(contextId in this._widgets)) {
+      this._widgets[contextId] = [];
+    }
+    this._widgets[contextId].push(parent);
+    let context = this._contextManager.getContext(contextId);
+    let model = this._contextManager.getModel(contextId);
+    // Create the child widget using the factory.
+    let child = wFactoryEx.factory.createNew(model, context, kernel);
+    this._attachChild(parent, child);
+    Private.nameProperty.set(parent, widgetName);
+    Private.contextProperty.set(parent, contextId);
+    installMessageFilter(parent, this);
+  }
+
+  /**
+   * Attach a child widget to a parent container.
+   */
+  private _attachChild(parent: Widget, child: Widget) {
+    parent.layout = new PanelLayout();
+    parent.title.closable = true;
+    parent.title.text = child.title.text;
+    parent.title.icon = child.title.icon;
+    parent.title.className = child.title.className;
+    // Mirror the parent title based on the child.
+    child.title.changed.connect(() => {
+      child.parent.title.text = child.title.text;
+      child.parent.title.icon = child.title.icon;
+      child.parent.title.className = child.title.className;
+    });
+    // Add the child widget to the parent widget.
+    (parent.layout as PanelLayout).addChild(child);
+  }
+
+  /**
+   * Ask the user whether to close an unsaved file.
+   */
+  private _maybeClose(widget: Widget, dirty: boolean): Promise<boolean> {
+    if (!dirty) {
+      return Promise.resolve(true);
+    }
+    let host = widget.isAttached ? widget.node : document.body;
+    return showDialog({
+      title: 'Close without saving?',
+      body: `File "${widget.title.text}" has unsaved changes, close without saving?`,
+      host
+    }).then(value => {
+      if (value && value.text === 'OK') {
+        return true;
+      }
+      return false;
+    });
+  }
+
+  /**
+   * Clean up the data associated with a widget.
+   */
+  private _cleanupWidget(widget: Widget): void {
+    // Remove the widget from our internal storage.
+    let id = Private.contextProperty.get(widget);
+    let index = this._widgets[id].indexOf(widget);
+    this._widgets[id] = this._widgets[id].splice(index, 1);
+    this._closeGuard = true;
+    // If this is the last widget in that context, remove the context.
+    if (!this._widgets[id]) {
+      let session = this._contextManager.removeContext(id);
+      if (session) {
+        // TODO: show a dialog asking whether to shut down the kernel.
+        widget.close();
+        widget.dispose();
+      }
+    } else {
+      widget.close();
```

Comment:
Are close-request messages triggered synchronously? If not, there is a race condition here where `this._closeGuard` is true, but checked in the message filter after possibly other stuff happening in the interim.